rclpy/test/test_clock.py

```diff
@@ -211,6 +216,137 @@ def test_triggered_clock_change_callbacks(self):
         handler3.unregister()
 
 
+@pytest.fixture()
+def default_context():
+    rclpy.init()
+    yield get_default_context()
+    rclpy.shutdown()
+
+
+@pytest.fixture()
+def non_default_context():
+    context = Context()
+    context.init()
+    yield context
+    context.try_shutdown()
+
+
+def test_sleep_until_mismatched_clock_type(default_context):
+    clock = Clock(clock_type=ClockType.SYSTEM_TIME)
+    with pytest.raises(ValueError, match='.*clock type does not match.*'):
+        clock.sleep_until(Time(clock_type=ClockType.STEADY_TIME))
+
+
+def test_sleep_until_non_default_context(non_default_context):
+    clock = Clock()
+    assert clock.sleep_until(clock.now() + Duration(seconds=0.1), context=non_default_context)
+
+
+def test_sleep_until_invalid_context():
+    clock = Clock()
+    with pytest.raises(NotInitializedException):
+        clock.sleep_until(clock.now() + Duration(seconds=0.1), context=Context())
+
+
+@pytest.mark.parametrize(
+    'clock_type', (ClockType.SYSTEM_TIME, ClockType.STEADY_TIME, ClockType.ROS_TIME))
+def test_sleep_until_basic(default_context, clock_type):
+    clock = Clock(clock_type=clock_type)
+    sleep_duration = Duration(seconds=0.1)
+    start = clock.now()
+    assert clock.sleep_until(clock.now() + sleep_duration)
+    stop = clock.now()
+    assert stop - start >= sleep_duration
+
+
+@pytest.mark.parametrize(
+    'clock_type', (ClockType.SYSTEM_TIME, ClockType.STEADY_TIME, ClockType.ROS_TIME))
+def test_sleep_until_time_in_past(default_context, clock_type):
+    clock = Clock(clock_type=clock_type)
+    sleep_duration = Duration(seconds=-1)
+    start = clock.now()
+    assert clock.sleep_until(clock.now() + sleep_duration)
+    stop = clock.now()
+    a_small_amount_of_time = Duration(seconds=0.01)
+    assert stop - start < a_small_amount_of_time
+
+
+@pytest.mark.parametrize('ros_time_enabled', (True, False))
+def test_sleep_until_ros_time_toggled(default_context, ros_time_enabled):
+    clock = ROSClock()
+    clock._set_ros_time_is_active(not ros_time_enabled)
+
+    retval = None
+
+    def run():
+        nonlocal retval
+        retval = clock.sleep_until(clock.now() + Duration(seconds=10))
+
+    t = threading.Thread(target=run, daemon=True)
+    t.start()
+
+    # wait for thread to get inside sleep_until call
+    time.sleep(0.2)
+
+    clock._set_ros_time_is_active(ros_time_enabled)
+
+    # wait for thread to exit
+    time.sleep(0.2)
```

Comment:
If we *don't* daemonize the thread, then we could do a `join` on the thread here instead, which at least gets rid of one sleep in the tests.  What do you think?

Reply:
`join()`ing the threads to avoid the second sleep in a8ddedf